### DISCUSSION TITLE
Resource fetching inside custom signal fails

### DISCUSSION THREAD

**Discussion by HavocNow** [solidjs/solid#1858] [Q&A]
Hi, 
I'm currentry trying to figure out how custom signals work in SolidJS. In my example app I need to fetch a list of meals in multiple components. For this I create a resource that uses a fetch function:
```JSX  const {token} = createAuth;
  const  fetchMeals= async () => {
    const result = await fetch(`http://localhost:8090/meals`,{headers: { Authorization: `Bearer ${token()}`}})
    const meals = (await result.json()).meals;
    return meals;
  }
  const [meals] = createResource(fetchMeals)
  ```

If I create the resource inside my component directly, everthing works as expected. Now I try to move the resource and the fetching logic into a custom signal:
#### createMeals.ts
```JSX 
import { createResource, createRoot } from "solid-js";
import createAuth from "./createAuth";

const createMeals = () => {
    const { token } = createAuth
    const fetchMeals = async () => {
        const result = await fetch(`http://localhost:8090/meals`, { headers: { Authorization: `Bearer ${token()}` } })
        const meals = (await result.json()).meals;
        return meals;
    }

    const [meals] = createResource(fetchMeals)
    return [meals]
}

export default createRoot(createMeals)
```

#### Component that uses the resource
```JSX
...
const [meals] = createMeals
...
```

At this point the fetching fails
![grafik](https://github.com/solidjs/solid/assets/7037551/ff2a111c-9429-4443-9d80-bc8d9671a1d8)

I have noticed that in some cases the data is fetched correctly by chance, so I assume it is some issue with the async fetch functions, but I don't have any solid ideas what the issue could be. Am I using resources / signals wrong or is it something else?

**Comment by snnsnn** [comment]
The terms you choose to explain your problem is not correct, hence can not describe it accurately.

There are no custom signals in solid JS, signal are signals. I believe what you mean was a custom hook: A custom function that returns a signal or an effect or other stuff that is part of core module. That is also misnomer because a hook is a callback that is called at certain point in time when executing certain functions. In this sense the only hook Solid has is `onDispose` where the registered callback is called when element is disposed. `onMount` is not a hook but a wrapper to create a non-reactive effect. That naming convention may make sense in React's context because React had component lifecycle and was/is calling certain function at certain point in time when running components but for Solid it is inaccurate and totally misleading. What people call a hook is just a function that creates a signal or effect or some other functions.

That being said, you are using Solid's API in a wrong way. For example, why would you call createMeals with createRoot. It does not make sense. You should take your time to understand the core concept and the API around them.

**Reply by HavocNow** [reply]
First of all, thanks for your detailed answer. You're right about the inaccuracy regarding custom hooks, what I tried to implement is basically what a custom hook in React would do. 
I tried to orient myself along this tutorial (https://www.solidjs.com/tutorial/stores_nocontext?solved) where something similar is done. To summarize, I would like to use a resource as a global state (for now without context just as an exercise). I will have another look at the docs and try to gain a deeper understanding of the underlying concepts, but in this case I tried to replicate something that is part of the basic tutorial. Could you maybe point me to a specific page / topic / concept so I can find out why the tutorial page works for signals but not for resources?

**Comment by HavocNow** [comment]
I managed to make it work by creating a simple signal and trigger the fetching with an effect:
```JSX
import { createEffect, createRoot, createSignal } from "solid-js";
import createAuth from "./createAuth";
import { AllMealsDto } from "../dto/dto";


const createMeals = () => {
    const { token } = createAuth
    const [meals, setMeals] = createSignal<Array<AllMealsDto>>([])
    const fetchMeals = async () => {
        const result = await fetch(`http://localhost:8090/meals`, { headers: { Authorization: `Bearer ${token()}` } })
        const meals = (await result.json()).meals;
        return meals;
    }

    createEffect(async () => {
        const result = await fetchMeals();
        setMeals(result);
    })

    return [meals]
}

export default createRoot(createMeals)
```
I always assumed that createResource is a shorthand for doing something similar (plus some extras) and can be user in the same way as a signal. Apparently that's not the case, so I'll try to figure out what the exact differences / limitations are.

**Reply by snnsnn** [reply]
`createResource` expect you to provide a fetcher function and returns an array with a signal and actions object:

```ts
const fetchMeals = async () => {
  const result = await fetch(`http://localhost:8090/meals`, { headers: { Authorization: `Bearer ${token()}` } })
  return  (await result.json()).meals;
}

const [data, actions] = createResource(fetchMeals);
```

You can pass query parameters through a signal:

```ts
const [data, actions] = createResource(queryParams(), fetchMeals);
```

You can wrap inside a function if you like, it will work the same way however it seems totally unnecessary abstraction, it is better to have components do the abstraction:

```ts
function createMeals() {
  const fetchMeals = async () => {
    const result = await fetch(`http://localhost:8090/meals`, { headers: { Authorization: `Bearer ${token()}` } })
    return  (await result.json()).meals;
  }

  return  createResource(fetchMeals);
}
```

The data is a signal already, you don't need to create another signal from it. `createResource` does the fetching, you don't need an effect to fetch the resource.

You can render a resource directly using Switch/Match components or use a Suspense. 

```tsx
const fetchMeals = async () => {
  const result = await fetch(`http://localhost:8090/meals`, { headers: { Authorization: `Bearer ${token()}` } })
  return  (await result.json()).meals;
}

const [data, actions] = createResource(fetchMeals);

<Switch>
  <Match when={data.state === 'pending'></Match>
  <Match when={data.state === 'errored'></Match>
  <Match when={data.state === 'ready'></Match>
</Switch>
```

`createRoot` is used to create a detached element, element that is not auto disposed but needs manual handling. If you don't know what it is, you should definitely stay away.

There are numerous questions on createResource, createRoot and all others API on both Solid's github repo and on stackoverflow. You can run a research and take a look at them:

https://github.com/solidjs/solid/discussions/1806
https://github.com/search?q=repo%3Asolidjs%2Fsolid+createResource&type=discussions
https://stackoverflow.com/search?q=%5Bsolid-js%5D+createResource

Please always do a thorough search before opening issues.
